cmd/util/ledger/migrations/change_contract_code_migration.go

```diff
@@ -1,175 +1,46 @@
-	defer d.mutex.RUnlock()
-
-	if len(d.contracts) > 0 {
-		var sb strings.Builder
-		sb.WriteString("failed to find all contract registers that need to be changed:\n")
-		for address, contracts := range d.contracts {
-			_, _ = fmt.Fprintf(&sb, "- address: %s\n", address)
-			for registerID := range contracts {
-				_, _ = fmt.Fprintf(&sb, "  - %s\n", flow.RegisterIDContractName(registerID))
-			}
-		}
-		return fmt.Errorf(sb.String())
-	}
-
-	return nil
-}
-
-func (d *ChangeContractCodeMigration) InitMigration(
-	log zerolog.Logger,
-	_ []*ledger.Payload,
-	_ int,
-) error {
-	d.log = log.
-		With().
-		Str("migration", "ChangeContractCodeMigration").
-		Logger()
-
-	return nil
-}
-
-func (d *ChangeContractCodeMigration) MigrateAccount(
-	_ context.Context,
-	address common.Address,
-	payloads []*ledger.Payload,
-) ([]*ledger.Payload, error) {
-
-	contracts, ok := (func() (map[flow.RegisterID]string, bool) {
-		d.mutex.Lock()
-		defer d.mutex.Unlock()
-
-		contracts, ok := d.contracts[address]
-
-		// remove address from set of addresses
-		// to keep track of which addresses are left to change
-		delete(d.contracts, address)
-
-		return contracts, ok
-	})()
-
-	if !ok {
-		// no contracts to change on this address
-		return payloads, nil
-	}
-
-	for payloadIndex, payload := range payloads {
-		key, err := payload.Key()
-		if err != nil {
-			return nil, err
-		}
-
-		registerID, err := convert.LedgerKeyToRegisterID(key)
-		if err != nil {
-			return nil, err
-		}
-
-		newContract, ok := contracts[registerID]
-		if !ok {
-			// not a contract register, or
-			// not interested in this contract
-			continue
-		}
-
-		// change contract code
-		payloads[payloadIndex] = ledger.NewPayload(
-			key,
-			[]byte(newContract),
-		)
-
-		// TODO: maybe log diff between old and new
-
-		// remove contract from list of contracts to change
-		// to keep track of which contracts are left to change
-		delete(contracts, registerID)
-	}
-
-	if len(contracts) > 0 {
-		var sb strings.Builder
-		_, _ = fmt.Fprintf(&sb, "failed to find all contract registers that need to be changed for address %s:\n", address)
-		for registerID := range contracts {
-			_, _ = fmt.Fprintf(&sb, "- %s\n", flow.RegisterIDContractName(registerID))
-		}
-		return nil, fmt.Errorf(sb.String())
-	}
-
-	return payloads, nil
-}
-
-func (d *ChangeContractCodeMigration) RegisterContractChange(
-	address common.Address,
-	contractName string,
-	newContractCode string,
-) (
-	previousNewContractCode string,
-) {
-	d.mutex.Lock()
-	defer d.mutex.Unlock()
-
-	if d.contracts == nil {
-		d.contracts = map[common.Address]map[flow.RegisterID]string{}
-	}
-
-	if _, ok := d.contracts[address]; !ok {
-		d.contracts[address] = map[flow.RegisterID]string{}
+func NewChangeContractCodeMigration(chainID flow.ChainID) *ChangeContractCodeMigration {
+	return &ChangeContractCodeMigration{
+		StagedContractsMigration: NewStagedContractsMigration(chainID).
+			// TODO:
+			//WithContractUpdateValidation().
```

Comment:
> Does this mean system contract updates are now going to be checked by the updated contract update checker, like other staged contract updates, and forced to pass?

@turbolent It is not - buy default. We can enable it here if we want.

Reply:
How can we disable it for system contracts and enable it for user contracts?

Reply:
It is also disabled in this PR. Enabled it in a follow-up PR: https://github.com/onflow/flow-go/pull/5415/files#diff-efc0583bcfd2aa19725211a31ed4ceed33b7d0ab5228c4ab533acf0d7f474d63R201-R202.

Reply:
I can do that change in this PR itself if that's better

Reply:
Nice! So with https://github.com/onflow/flow-go/pull/5415/files#diff-efc0583bcfd2aa19725211a31ed4ceed33b7d0ab5228c4ab533acf0d7f474d63R201-R202, now all contract updates, both "system contracts" and "user contracts", will get validated?

Reply:
so basically, we can enable/disable it in the corresponding constructor `NewCadence1ContractsMigrations` and `NewChangeContractCodeMigration`

Reply:
Great! Thank you for the explanation

Reply:
> now all contract updates, both "system contracts" and "user contracts", will get validated?

only the user contracts get validated for now. Validation for system contracts still remains disabled.

Reply:
Sorry, the github isn't showing (your) comments until I refresh and I replied without refreshing 😅